src/java/org/apache/cassandra/utils/units/Units.java

```diff
@@ -0,0 +1,233 @@
+     * <p>
+     * For instance, {@code formatValue(4693234L) == "4,693,234"}
+     *
+     * @param value the value to format.
+     * @return a more human readable representation of {@code value}.
+     */
+    static String formatValue(long value)
+    {
+        String v = Long.toString(value);
+        int l = v.length();
+
+        int digits = value < 0 ? l - 1 : l;
+        int commaCount = commaCount(digits);
+        if (commaCount == 0)
+            return v;
+
+        char[] chars = new char[l + commaCount];
+        int signShift = value < 0 ? 1 : 0;
+        // Copy all digits to their proper position, which is their original position shifted to right from how
+        // many comma will be on their left (which itself is the total number of comma minus the one remaining to add
+        // on the right).
+        for (int i = 0; i < digits; i++)
+            chars[signShift + i + (commaCount - commaCount(digits - i))] = v.charAt(signShift + i);
+        // Then add all comma (which, starting from the left, are every 4 characters).
+        for (int i = 1; i <= commaCount; i++)
+            chars[chars.length - (4 * i)] = ',';
+        // Lastly, add the minus sign on negative values
+        if (value < 0)
+            chars[0] = '-';
+        return new String(chars);
+    }
+
+    /**
+     * The number of comma to use to format {@code digits} digit using ',' on every thousands.
+     */
+    private static int commaCount(int digits)
+    {
+        return (digits - 1) / 3;
+    }
+
+    private static <E extends Enum<E>> String toString(long value,
+                                                       E unit,
+                                                       Class<E> klass,
+                                                       ToLongFunction<E> scaleFct,
+                                                       Function<E, String> symbolFct)
+    {
+        E[] enumVals = klass.getEnumConstants();
+
+        long v = value;
+        int i = unit.ordinal();
+        long remainder = 0;
+        // The scale is how much we need to go from unit to the next one
+        long scale = scaleFct.applyAsLong(unit);
+
+        while (i < enumVals.length - 1 && v >= scale)
+        {
+            remainder = v % scale;
+            v = v / scale;
+            unit = enumVals[++i];
+            scale = scaleFct.applyAsLong(unit);
+        }
+
+        // If the value is small (<10), include one decimal so the precision is not too truncated. Otherwise, don't
+        // bother, it's less relevant.
+        if (v >= 10 || remainder == 0)
+            return fmt(v, unit, symbolFct);
+
+        // Note that scale is the scale of the current unit, but remainder relates to the previous unit. Also not that
+        // can only get here is remainder != 0 so we know accessing the previous unit is legit
+        long prevScale = scaleFct.applyAsLong(enumVals[i - 1]);
```

Comment:
Could `i == 0` here?

Reply:
it can't according to the doc in the line above